python_modules/dagster/dagster_tests/model_tests/test_dagster_model.py

```diff
@@ -1,41 +1,82 @@
+from typing import Optional
+
 import pytest
-from dagster._model import DagsterModel
+from dagster._check import CheckError
+from dagster._model import DagsterModel, copy, dagster_model, dagster_model_with_new
+from dagster._model.decorator import SerdesOptions
+from dagster._serdes.serdes import deserialize_value, serialize_value
 from dagster._utils.cached_method import CACHED_METHOD_CACHE_FIELD, cached_method
 from pydantic import ValidationError
 
 
-def test_runtime_typecheck():
+def test_runtime_typecheck() -> None:
     class MyClass(DagsterModel):
         foo: str
         bar: int
 
     with pytest.raises(ValidationError):
-        MyClass(foo="fdsjk", bar="fdslk")
+        MyClass(foo="fdsjk", bar="fdslk")  # type: ignore # good job type checker
+
+    @dagster_model
+    class MyClass2:
+        foo: str
+        bar: int
+
+    with pytest.raises(CheckError):
+        MyClass2(foo="fdsjk", bar="fdslk")  # type: ignore # good job type checker
 
 
-def test_override_constructor_in_subclass():
+def test_override_constructor_in_subclass() -> None:
     class MyClass(DagsterModel):
         foo: str
         bar: int
 
         def __init__(self, foo: str, bar: int):
             super().__init__(foo=foo, bar=bar)
 
-    MyClass(foo="fdsjk", bar=4)
+    assert MyClass(foo="fdsjk", bar=4)
+
+    @dagster_model
+    class MyClass2:
+        foo: str
+        bar: int
+
+        def __new__(cls, foo: str, bar: int):
+            return super().__new__(
+                cls,
+                foo=foo,  # type: ignore
+                bar=bar,  # type: ignore
+            )
+
+    assert MyClass2(foo="fdsjk", bar=4)
 
 
-def test_override_constructor_in_subclass_different_arg_names():
+def test_override_constructor_in_subclass_different_arg_names() -> None:
     class MyClass(DagsterModel):
         foo: str
         bar: int
 
         def __init__(self, fooarg: str, bararg: int):
             super().__init__(foo=fooarg, bar=bararg)
 
-    MyClass(fooarg="fdsjk", bararg=4)
+    assert MyClass(fooarg="fdsjk", bararg=4)
 
+    @dagster_model_with_new
+    class MyClass2:
+        foo: str
+        bar: int
+
+        def __new__(cls, fooarg: str, bararg: int):
+            return super().__new__(
+                cls,
+                foo=fooarg,  # type: ignore
+                bar=bararg,  # type: ignore
```

Comment:
[X] type system doesn't like the super().__new__ args since it doesn't know about them